graph/nodes_edges.go

```diff
@@ -56,28 +56,29 @@ type NodeSlicer interface {
 
 // NodesOf returns it.Len() nodes from it. If it is a NodeSlicer, the NodeSlice method
 // is used to obtain the nodes. It is safe to pass a nil Nodes to NodesOf.
-//
-// If the Nodes has an indeterminate length, NodesOf will panic.
 func NodesOf(it Nodes) []Node {
 	if it == nil {
 		return nil
 	}
-	len := it.Len()
+	n := it.Len()
 	switch {
-	case len == 0:
+	case n == 0:
 		return nil
-	case len < 0:
-		panic("graph: called NodesOf on indeterminate iterator")
+	case n < 0:
+		n = 0
 	}
 	switch it := it.(type) {
 	case NodeSlicer:
 		return it.NodeSlice()
 	}
-	n := make([]Node, 0, len)
+	nodes := make([]Node, 0, n)
 	for it.Next() {
-		n = append(n, it.Node())
+		nodes = append(nodes, it.Node())
 	}
-	return n
+	if len(nodes) == 0 {
+		return nil
+	}
```

Comment:
Why not move it before the for loop? Can n == 0 and it.Next() == true?

Reply:
We don't know the length of the slice in the case of an indeterminate iterator until after the iterator has completed.